pkg/jobs/job_scheduler.go

```diff
@@ -70,65 +70,66 @@ func newJobScheduler(
 
 const allSchedules = 0
 
-// getFindSchedulesStatement returns SQL statement used for finding
-// scheduled jobs that should be started.
-func getFindSchedulesStatement(env scheduledjobs.JobSchedulerEnv, maxSchedules int64) string {
-	limitClause := ""
-	if maxSchedules != allSchedules {
-		limitClause = fmt.Sprintf("LIMIT %d", maxSchedules)
-	}
+var errScheduleNotRunnable = errors.New("schedule not runnable")
 
-	return fmt.Sprintf(
-		`
-SELECT
-  (SELECT count(*)
-   FROM %s J
-   WHERE
-      J.created_by_type = '%s' AND J.created_by_id = S.schedule_id AND
-      J.status NOT IN ('%s', '%s', '%s', '%s')
-  ) AS num_running, S.*
-FROM %s S
-WHERE next_run < %s
-ORDER BY random()
-%s
-FOR UPDATE`, env.SystemJobsTableName(), CreatedByScheduledJobs,
-		StatusSucceeded, StatusCanceled, StatusFailed, StatusRevertFailed,
-		env.ScheduledJobsTableName(), env.NowExpr(), limitClause)
-}
-
-// unmarshalScheduledJob is a helper to deserialize a row returned by
-// getFindSchedulesStatement() into a ScheduledJob
-func (s *jobScheduler) unmarshalScheduledJob(
-	row []tree.Datum, cols []colinfo.ResultColumn,
-) (*ScheduledJob, int64, error) {
-	j := NewScheduledJob(s.env)
-	if err := j.InitFromDatums(row[1:], cols[1:]); err != nil {
-		return nil, 0, err
+// loadCandidateScheduleForExecution looks up and locks candidate schedule for execution.
+// The schedule is locked via FOR UPDATE clause to ensure that only this scheduler can modify it.
+// If schedule cannot execute, a errScheduleNotRunnable error is returned.
+func loadCandidateScheduleForExecution(
+	ctx context.Context,
+	scheduleID int64,
+	env scheduledjobs.JobSchedulerEnv,
+	ie sqlutil.InternalExecutor,
+	txn *kv.Txn,
+) (*ScheduledJob, error) {
+	lookupStmt := fmt.Sprintf(
+		"SELECT * FROM %s WHERE schedule_id=%d AND next_run < %s FOR UPDATE",
+		env.ScheduledJobsTableName(), scheduleID, env.NowExpr())
+	row, cols, err := ie.QueryRowExWithCols(
+		ctx, "find-scheduled-jobs-exec",
+		txn,
+		sessiondata.InternalExecutorOverride{User: security.RootUserName()},
+		lookupStmt)
+	if err != nil {
+		return nil, err
 	}
 
-	if n, ok := row[0].(*tree.DInt); ok {
-		return j, int64(*n), nil
+	if row == nil {
+		return nil, errScheduleNotRunnable
```

Comment:
Should we add (wrap) a little information here, such as the `scheduleID`? Just in case we this happens more frequently than we expect.